manta-crypto/src/arkworks/glv/mod.rs

```diff
@@ -0,0 +1,343 @@
+{
+    #[inline]
+    fn x(&self) -> &Self::BaseField {
+        &self.x
+    }
+
+    #[inline]
+    fn y(&self) -> &Self::BaseField {
+        &self.y
+    }
+
+    #[inline]
+    fn from_xy_unchecked(x: Self::BaseField, y: Self::BaseField) -> Self {
+        Self::new(x, y, false)
+    }
+}
+
+/// Given a scalar `k` and basis vectors `v` and `u` finds integer scalars `k1` and `k2`,
+/// so that `(k, 0)` is close to `k1v + k2u`.
+#[inline]
+pub fn decompose_scalar<F>(k: &F, v: (&BigInt, &BigInt), u: (&BigInt, &BigInt)) -> (BigInt, BigInt)
+where
+    F: PrimeField,
+{
+    // TODO: Check the bit length of BitInt v.s. BaseField or ScalarField. See if there is any overflow issue.
+    let k = BigInt::from_bytes_be(Sign::Plus, &k.into_repr().to_bytes_be());
+    let q1 = (u.1 * &k) / ((v.0 * u.1) - (v.1 * u.0));
+    let q2 = (-v.1 * &k) / ((v.0 * u.1) - (v.1 * u.0));
+    let k1 = k - &q1 * v.0 - &q2 * u.0;
+    let k2 = 0 - q1 * v.1 - q2 * u.1;
+    (k1, k2)
+}
+
+#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+/// GLV Parameters
+pub struct GLVParameters<C>
+where
+    C: AffineCurve,
+{
+    /// Endomorphism-defining Element
+    pub beta: C::BaseField,
+
+    /// Basis Element
+    pub base_v1: (BigInt, BigInt),
+
+    /// Basis Element
+    pub base_v2: (BigInt, BigInt),
+}
+
+impl<C> GLVParameters<C>
+where
+    C: AffineCurve,
+{
+    /// Creates a new instance of [`GLVParameters`] without checking that `base_v1` and `base_v2` form a
+    /// basis of the kernel of the function defined by the root of the characteristic polynomial
+    /// of the endomorphism associated with `beta`.
+    #[inline]
+    pub fn new_unchecked(
+        beta: C::BaseField,
+        base_v1: (BigInt, BigInt),
+        base_v2: (BigInt, BigInt),
+    ) -> Self {
+        Self {
+            beta,
+            base_v1,
+            base_v2,
+        }
+    }
+
+    /// Creates a new instance of [`GLVParameters`] from the curve parameters.
+    #[inline]
+    pub fn new<M>() -> Self
+    where
+        C: HasGLV<M>,
+    {
+        C::glv_parameters()
+    }
+
+    /// Returns a reference to `beta`.
+    pub fn beta(&self) -> &C::BaseField {
+        &self.beta
+    }
+
+    /// Returns `beta`, dropping `self`.
+    pub fn into_beta(self) -> C::BaseField {
+        self.beta
+    }
+
+    /// Returns a reference to the basis elements.
+    pub fn basis(&self) -> ((&BigInt, &BigInt), (&BigInt, &BigInt)) {
+        (
+            (&self.base_v1.0, &self.base_v1.1),
+            (&self.base_v2.0, &self.base_v2.1),
+        )
+    }
+
+    /// Generates scalars and points for the simultaneous multiple
+    /// point multiplication.
+    #[inline]
+    fn scalars_and_points(
+        &self,
+        point: &C,
+        scalar: &C::ScalarField,
+    ) -> (Vec<bool>, Vec<bool>, C::Projective, C::Projective)
+    where
+        C: AffineCurveExt,
+    {
+        // TODO: make sure both output vectors have the same length. I guess into_repr and
```

Comment:
TODO

Reply:
just checked the docs, it seems that they do indeed. TODO removed.